In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [2]:
# Crear un dataframe
df_numericos = pd.read_csv('CSV/numericos.csv', index_col=0) 

In [3]:
df_numericos.head(10)

age    dailyrate  distancefromhome  education  employeecount  \
0   51  2015.722222                 6          3              1   
1   52  2063.388889                 1          4              1   
2   42  1984.253968                 4          2              1   
3   47  1771.404762                 2          4              1   
4   46  1582.771346                 3          3              1   
5   48  1771.920635                22          3              1   
6   59  1032.487286                25          3              1   
7   42   556.256661                 1          1              1   
8   41  1712.182540                 2          5              1   
9   41  1973.984127                 9          3              1   

   employeenumber  environmentsatisfaction  hourlyrate  jobinvolvement  \
0               1                        1         NaN               3   
1               2                        3         NaN               2   
2               3                        3         NaN               3   
3               4                        1         NaN               3   
4               5                        1         NaN               4   
5               6                        4         NaN               3   
6               7                        1         NaN               3   
7               8                        2   69.532083               3   
8               9                        2         NaN               3   
9              10                        1         NaN               3   

   joblevel  jobsatisfaction  monthlyincome  monthlyrate  numcompaniesworked  \
0         5                3       16280.83     42330.17                   7   
1         5                3            NaN     43331.17                   0   
2         5                4            NaN     41669.33                   1   
3         4                3       14307.50     37199.50                   3   
4         4                1       12783.92     33238.20                   2   
5         4                4       14311.67     37210.33                   3   
6         3                1        8339.32     21682.23                   7   
7         2                3            NaN     11681.39                   1   
8         4                1       13829.17     35955.83                   7   
9         5                3       15943.72     41453.67                   2   

   percentsalaryhike  performancerating  relationshipsatisfaction  \
0                 13                3.0                         3   
1                 14                3.0                         1   
2                 11                3.0                         4   
3                 19                3.0                         2   
4                 12                3.0                         4   
5                 11                3.0                         2   
6                 11                3.0                         4   
7                 25                4.0                         3   
8                 16                3.0                         2   
9                 17                3.0                         2   

   stockoptionlevel  totalworkingyears  trainingtimeslastyear  \
0                 0                NaN                      5   
1                 1               34.0                      5   
2                 0               22.0                      3   
3                 2                NaN                      2   
4                 1                NaN                      5   
5                 1                NaN                      3   
6                 0               28.0                      3   
7                 0               20.0                      3   
8                 1               22.0                      2   
9                 1               21.0                      2   

   worklifebalance  yearsatcompany  yearsincurrentrole  \
0              3.0              20 

In [4]:
df_numericos.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1678 entries, 0 to 1677
Data columns (total 29 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       1678 non-null   int64  
 1   dailyrate                 1678 non-null   float64
 2   distancefromhome          1678 non-null   int64  
 3   education                 1678 non-null   int64  
 4   employeecount             1678 non-null   int64  
 5   employeenumber            1678 non-null   int64  
 6   environmentsatisfaction   1678 non-null   int64  
 7   hourlyrate                411 non-null    float64
 8   jobinvolvement            1678 non-null   int64  
 9   joblevel                  1678 non-null   int64  
 10  jobsatisfaction           1678 non-null   int64  
 11  monthlyincome             1189 non-null   float64
 12  monthlyrate               1678 non-null   float64
 13  numcompaniesworked        1678 non-null   int64  
 14  percentsalary

In [5]:
df_numericos.columns

Index(['age', 'dailyrate', 'distancefromhome', 'education', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'hourlyrate',
       'jobinvolvement', 'joblevel', 'jobsatisfaction', 'monthlyincome',
       'monthlyrate', 'numcompaniesworked', 'percentsalaryhike',
       'performancerating', 'relationshipsatisfaction', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'numberchildren'],
      dtype='object')

## Información columnas y posibles cambios: 

#### 1) Columna 'age': 

Info: 
- Descripción: La edad del empleado
- Datatype: int64
- Duplicados: N/A (puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

#### 2) Columna 'dailyrate': 

Info: 
- Descripción: La tarifa diaria del empleado.
- Datatype: float64 (no redondeado) 
- Duplicados: N/A (el salario diario puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- Mover el decimal una posición a la izquierda (EJ: fila 1677, el hourly rate es 69.532083 y el dailyrate es 556.256661. Si movemos el decimal y dividimos 55.6256661 por 8, da 6.9532083)
- Redondear a dos decimales? 

#### 3) Columna 'distancefromhome': 

Info: 
- Descripción: La distancia desde el hogar del empleado hasta su lugar de trabajo.
- Datatype: int64
- Duplicados: N/A (la distancia puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- Hay distancias negativas (202). Verificar si es posible cambiar a positivo (los km no pueden ser negativos)

-- Contar cuántos valores negativos hay en la columna "distancefromhome": 
negativos = (df["distancefromhome"] < 0).sum()
print(f"Número de valores negativos en 'distancefromhome': {negativos}")-- 

#### 4) Columna 'education': 

Info: 
- Descripción: Nivel de educación del empleado (generalmente en una escala del 1 al 5).
- Datatype: int64
- Duplicados: N/A (la education puede repetirse por empleado)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: No en esta, pero quizas si en la columna educationfield por medio del education level. 
- Columna similar: educationfield?
- Mantener como columna numérica? Si
- Unique: [3, 4, 2, 1, 5]
- Value counts: 
3    649
4    461
2    322
1    186
5     60

Observaciones / Posibles cambios:
- Ver si tiene relación con la columna categorica 'educationfield'. Ver en comprobaciones el código de contingencia: 

#Ver contingencia entre las dos tablas: (para verificar si hay relacion entre education y field) 
contingency_table = pd.crosstab(df['education'], df['educationfield'])
contingency_table

#### 5) Columna 'employeecount': 

Info: 
- Descripción: Un contador que generalmente es 1 y se usa para contar empleados.
- Datatype: int64
- Duplicados: N/A (siempre es 1)
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: 1
- Value counts: N/A

Observaciones / Posibles cambios:
- Ver si es útil o podemos eliminarlo. 

#### 6) Columna 'employeenumber': 

Info: 
- Descripción: Un número de identificación único para el empleado.
- Datatype: int64
- Duplicados: 64
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: n/a (son muchos)
- Value_counts: n/a (son muchos)

Observaciones / Posibles cambios:
- Eliminar duplicados. 


#### 7) Columna 'environmentsatisfaction': 

Info: 
- Descripción: Nivel de satisfacción del empleado en relación con su entorno de trabajo. Con valores que estan comprendidos entre el 1 y el 4, siendo el 4 el nivel de máxima satisfacción. 
- Datatype: int64
- Duplicados: 0
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si
- Unique: *Deberia ser del 1 al 4, pero va del 1 al 49*
- Value_counts: (son muchos)

Observaciones / Posibles cambios:
- Ver si tiene que ver con el departamento o otra columna para modificar los datos. O si es posible que el 49 sea 4,9 o algo asi. 


#### 8) Columna hourlyrate: 

Info: 
- Descripción: La tarifa por hora del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 1267 / En caso se ser mayoría de nulos, puede completarse con otra columna?: (ver con el resto de las columnas salariales)
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
- ver si se puede completar con la columna de monthly rate (ver tmb lo de full o part time)


#### 9) Columna jobinvolvement: 

Info: 
- Descripción: Nivel de implicación del empleado en su trabajo.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 10) Columna joblevel: 

Info: 
- Descripción: Nivel jerárquico del empleado en la empresa.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 11) Columna jobsatisfaction: 

Info: 
- Descripción: Nivel de satisfacción del empleado con su trabajo.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: relationshipsatisfaction ?
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver si 1 es el peor o mejor nivel de satisfaccion. 

#### 12) Columna monthlyincome: 

Info: 
- Descripción: Ingresos mensuales del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 489 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A - ver si hay relacion con salary o sameasmonthlyincome en caso que haya mas info ahi 
- Columna similar: sameasmonthlyincome
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver relacion con sameasmonthlyincome o salary

#### 13) Columna monthlyrate: 

Info: 
- Descripción: Tasa mensual del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A - ver con hourlyrate
- Columna similar: sameasmonthlyincome
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver relacion con hourlyrate

#### 14) Columna numcompaniesworked: 

Info: 
- Descripción: Número de compañías en las que el empleado ha trabajado.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A


#### 15) Columna percentsalaryhike: 

Info: 
- Descripción: El porcentaje de aumento salarial del empleado.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 16) Columna performancerating: 

Info: 
- Descripción: Calificación de rendimiento del empleado
- Datatype: float64
- Duplicados: N/A
- Nulos: 200 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 17) Columna relationshipsatisfaction: 

Info: 
- Descripción:  Nivel de satisfacción en las relaciones interpersonales del empleado.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: jobsatisfaction?
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 18) Columna stockoptionlevel: 

Info: 
- Descripción:  Nivel de opciones de compra de acciones del empleado
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 19) Columna totalworkingyears: 

Info: 
- Descripción:  Total de años de experiencia laboral del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 549 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: yearsatcompany??
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 20) Columna trainingtimeslastyear: 

Info: 
- Descripción:  Número de veces que el empleado recibió capacitación el año pasado.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 21) Columna worklifebalance: 

Info: 
- Descripción:  Equilibrio entre trabajo y vida personal del empleado.
- Datatype: float64
- Duplicados: N/A
- Nulos: 114 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A

#### 22) Columna yearsatcompany: 

Info: 
- Descripción:  Años que el empleado ha trabajado en la empresa actual.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: totalworkingyears ?
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver dif entre totalworkingyears -si es de toda su vida- o si es similar a yearsatcompany


#### 23) Columna yearsincurrentrole: 

Info: 
- Descripción:  Años que el empleado ha estado en su puesto actual
- Datatype: float64
- Duplicados: N/A
- Nulos: 1643 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
gestionar nulos (son muchos)

#### 24) Columna yearssincelastpromotion: 

Info: 
- Descripción:  Años desde la última promoción del empleado.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A


#### 25) Columna yearswithcurrmanager: 

Info: 
- Descripción:  Años que el empleado ha estado bajo la supervisión del actual gerente.
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A 
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A


#### 26) Columna sameasmonthlyincome: 

Info: 
- Descripción:  Ingresos mensuales del empleado
- Datatype: float64
- Duplicados: N/A
- Nulos: 489 / En caso se ser mayoría de nulos, puede completarse con otra columna?: tiene los mismos  nulos que monthlyincome
- Columna similar: monthlyincome
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver si es necesaria (parece igual a monthly income)

#### 27) Columna datebirth: 

Info: 
- Descripción: Año de nacimiento del empleado (teniendo en cuenta que los datos fueron recogidos en el 2023)
- Datatype: int64
- Duplicados: N/A
- Nulos: 0 / En caso se ser mayoría de nulos, puede completarse con otra columna?: tiene los mismos  nulos que monthlyincome
- Columna similar: 
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
N/A


#### 28) Columna salary: 

Info: 
- Descripción:  salario de los empleados
- Datatype: float64
- Duplicados: N/A
- Nulos: 285 / En caso se ser mayoría de nulos, puede completarse con otra columna?: ver monthlyincome o monthlyrate a ver si hay relacion
- Columna similar: monthlyincome / monthlyrate * 12 ?
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
ver relacion con las otras dos de income / rate

#### 29) Columna numberchildren: 

Info: 
- Descripción:  numero de hijos
- Datatype: float64 - ver si deberia ser int 
- Duplicados: N/A
- Nulos: 1677 / En caso se ser mayoría de nulos, puede completarse con otra columna?: N/A
- Columna similar: N/A
- Mantener como columna numérica? Si

Observaciones / Posibles cambios:
gestion de nulos (la mayoria)

## Comprobaciones

In [14]:
df_numericos.columns

Index(['age', 'dailyrate', 'distancefromhome', 'education', 'employeecount',
       'employeenumber', 'environmentsatisfaction', 'hourlyrate',
       'jobinvolvement', 'joblevel', 'jobsatisfaction', 'monthlyincome',
       'monthlyrate', 'numcompaniesworked', 'percentsalaryhike',
       'performancerating', 'relationshipsatisfaction', 'stockoptionlevel',
       'totalworkingyears', 'trainingtimeslastyear', 'worklifebalance',
       'yearsatcompany', 'yearsincurrentrole', 'yearssincelastpromotion',
       'yearswithcurrmanager', 'sameasmonthlyincome', 'datebirth', 'salary',
       'numberchildren'],
      dtype='object')

In [134]:
# Contar cuántos valores negativos hay una columna
negativos1 = (df_numericos["numberchildren"] < 0).sum()
print(f"Número de valores negativos en 'numberchildren': {negativos1}")

Número de valores negativos en 'numberchildren': 0


In [139]:
df_numericos['numberchildren'].dtype

dtype('float64')

In [145]:
# Verificar duplicados
df_numericos['numberchildren'].duplicated().sum()


1677

In [144]:
#Ver nulos
df_numericos['numberchildren'].isnull().sum()

1678

In [143]:
df_numericos["numberchildren"].unique()

array([nan])

In [142]:
df_numericos['numberchildren'].value_counts()

Series([], Name: count, dtype: int64)